### PR TITLE
fixing due to comment

### PR DESCRIPTION
**Pull request by Maraljan**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
-from RPA.Browser.Selenium import Selenium
-from RPA.Excel.Files import Files
 import os
+from datetime import timedelta
```
Comment (Maraljan) [inline main.py:4]: You should put a blank line between each group of imports. (PEP8)

File: main.py [hunk 2]
```diff
@@ -10,51 +12,57 @@ def open_the_website(url):
     download_path = os.path.abspath("output")
     pref = {"download.default_directory": download_path}
     browser_lib.open_available_browser(url=url, preferences=pref)
+    browser_lib.set_download_directory(download_path)
+    # special method <set_download_directory> has no effect
```
Comment (Maraljan) [inline main.py]: Method **set_download_directory**  is used before the opening browser.

File: main.py [hunk 3]
```diff
@@ -63,29 +71,25 @@ def scrape_table():
             data_set.append(data.text)
```
Comment (Maraljan) [inline main.py:83]: It would be nice to collect links from the table here in this **for** (64 line). Less operation.

File: main.py [hunk 4]
```diff
@@ -94,3 +98,13 @@ def main():
 if __name__ == "__main__":
     main()
 
+# The task below positioned as bonus (optional).
```
Comment (Maraljan) [inline main.py]: Try to complete this task, especially you have already done half.

File: main.py [hunk 5]
```diff
@@ -94,3 +98,13 @@ def main():
...
+
+
+# 'Extract data from PDF. You need to get the data from Section A in each PDF.
+# Then compare the value "Name of this Investment" with the column "Investment Title",
+# and the value "Unique Investment Identifier (UII)" with the column "UII"'
+
+
+# It is not mentioned what exactly the program should do after comparing. What should happen if, for example,
+# the values in pdf and table are different?
```
Comment (Maraljan) [inline main.py]: You should display a message that the "Name of this Investment" value from the table is equal to the "Investment Title" value from the PDF, and also with the UII. For messages use builtin [logging ](https://docs.python.org/3/library/logging.html)or library [loguru](https://loguru.readthedocs.io/en/stable/).
#### Try to find an optimal way.

**Review by Maraljan — COMMENTED** [review]
### This file appears in the folder, why do we need it?
image

### Check if the number of agencies in the table matches the number on the site, because I think they are lost.
image

### If you have any questions, you can write your question under the comment.

File: main.py [hunk 6]
```diff
@@ -60,51 +70,84 @@ def scrape_table(headers):
...
+    element = 'id:business-case-pdf >> tag:a'
+
+    browser_lib.wait_until_element_is_visible(element, timeout=timedelta(seconds=20))
+    browser_lib.click_element(element)
+    browser_lib.assign_id_to_element(locator=element, id="USERFLAG")
+
+    condition = "return document.getElementById('USERFLAG').getAttribute('aria-busy') == \"false\""
+    browser_lib.wait_for_condition(condition=condition, timeout=timedelta(seconds=20))
+
+    time.sleep(3)  # wait till file is downloaded
```
Comment (Maraljan) [inline main.py]: you can use instead of 106-107 lines of code the method **wait_until_created** from FileSystem and you don't have to use **sleep**

File: main.py [hunk 7]
```diff
@@ -60,51 +70,84 @@ def scrape_table(headers):
...
+
+def compare_values(equalities, file):
+    text = pdf.get_text_from_pdf(file)[1]
+    investment = re.search(r"1\. Name of this Investment: (.*)2\.", text).group(1)
+    uii = re.search(r"2\. Unique Investment Identifier \(UII\): (.*)Section B", text).group(1)
+
+    if equalities[uii] == investment:
+        logging.info(uii, investment, "EQUAL")
+    else:
+        logging.info(uii, investment, "NOT EQUAL")
```
Comment (Maraljan) [inline main.py]: **Logs are not shown:**
image

File: main.py [hunk 8]
```diff
@@ -60,51 +70,84 @@ def scrape_table(headers):
...
     for link in links:
-        browser_lib.go_to(link)
-        element = '//*[@id="business-case-pdf"]/a'
-        browser_lib.wait_until_element_is_visible(element, timeout=timedelta(seconds=10))
-        browser_lib.click_element(element)
-        browser_lib.assign_id_to_element(locator=element, id="USERFLAG")
-        condition = "return document.getElementById('USERFLAG').getAttribute('aria-busy') == \"false\""
-        browser_lib.wait_for_condition(condition=condition, timeout=timedelta(seconds=10))
+        file = download_file(link)
+        compare_values(equalities, file)
```
Comment (Maraljan) [inline main.py]: So u compare **values from a pdf** with **all data from the table**, but you need to compare one to one (one pdf to one row from the table)

File: main.py [hunk 9]
```diff
@@ -24,22 +29,26 @@ def click_button():
...
     container_text = container_element.text.split('\n')
+
     result_list = list(zip(container_text[::4], container_text[2::4]))
     result_list.insert(0, ("Agency name", "Spending"))
+
     return result_list
 
 
-def write_to_excel(sheetname, bookname, content):
+def write_to_excel(sheetname, bookname, content, method):
```
Comment (Maraljan) [inline main.py]: why did you add the **method** argument? The previous solution is better and more optimal.

**Review by Maraljan — COMMENTED** [review]
Come on, there is not much left) Regex is very very nice. 

If something is not clear feel free to contact me.

File: main.py (hunk 9 above)
Comment (yakimart) [reply main.py]: The previous solution was only appending data to excel file. It means that for each new launching of the program, excel files contains data from previous launch and gets appended with new data. I add method option to make usage of function more flexible. Now function is beeing used to create book for agencies list (method "new") and to add sheet with spendings to it (method "append")

File: main.py (hunk 8 above)
Comment (yakimart) [reply main.py]: Dictionary equalities contains pairs where UII is key and investment is value. In other words, each row of table is represented in dictionary like key: value pair. Once pdf is downloaded, UII and investment from THIS pdf are beeing compared with value from dictionary with UII as a unique key

File: main.py (hunk 7 above)
Comment (yakimart) [reply main.py]: Depends on environment. Set it like warning instead of info for assuarance in every environment

File: main.py (hunk 6 above)
Comment (yakimart) [reply main.py]: Solved it

File: main.py (hunk 8 above)
Comment (Maraljan) [reply main.py]: Let's imagine a situation where we have two rows with the same UII, then you will have only **one** entry in the dictionary, in this case, you will download **two** PDFs, you will compare them with **one** row but u should with **two**.

File: main.py (hunk 9 above)
Comment (Maraljan) [reply main.py]: In Robocloud anyway, everything will be deleted after the script ends. You can just check if the table exists. If yes then add it else create a new one - similar to your first solution.

File: main.py (hunk 7 above)
Comment (Maraljan) [reply main.py]: this solution is okay. But in the future, u can set **logger level** to:

- debug 
- info 
- warning
- error

**Review by Maraljan — COMMENTED** [review]
**I wanted to know if everything works in Robocloud?**

**The number of agencies on the site and the number of  data in the agency table are not equal**
image

File: main.py (hunk 8 above)
Comment (yakimart) [reply main.py]: UII is a unique value, it can't be presented in two different rows (in this case)

File: main.py (hunk 9 above)
Comment (yakimart) [reply main.py]: Checked out to the previous function solution

**Comment by yakimart** [comment]
1. Amount of agencies in excel file. You showed two different screenshots with different results but code wasn't changed, it's odd.. I have always the same correct amount of agencies, in robocloud too.
2. Execution in robocloud terminates after 3 minutes because of cloud basic plan limitation for execution time, as I understood. All stages before (scraping agencies and investments, writing to excel file, downloading pdf and comparing values) are working properly.
3. I've tested my solution on my local robocorp environment. The only issue is denying for cleaning function that removes files from directory which are not excel or pdf files. Also robocloud creates log file, thats why I suggest not to remove any files from the output folder.

**Comment by Maraljan** [comment]
1. I specially launched a couple of times to check the number of agencies and each time they have a different result, try to run it several times and check. If everything is clear, then the problem is in me.
2. I understand that UII is unique but we should check **one** **row** to **one pdf** in case of any duplication.

**Comment by yakimart** [comment]
1. After a few runs, amounts of agencies in excel are the same and correct. 
image
2. Added solution for uii duplicates. 
3. What's next?
